### PR TITLE
Week 2 Step 5

### PR DESCRIPTION
**Pull request by anamatasanch**
+Add maven dependency to use datastore
+Add Comment class to store comments as objects
+Modify servlet to work with comment objects.
+Modify servlet to read and add comments from the datastore
+Modify script.js to show both the text and timestamp of the comment in the comments container

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.google.sps.servlets;
+
+public class Comment {
+  long id;
+  String text;
+  String timestamp;
+
+  public Comment(){
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Why do we need this constructor?

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.google.sps.servlets;
+
+public class Comment {
+  long id;
+  String text;
+  String timestamp;
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java]: I suggest to add the visibility, e.g. private final long id; The same for the other fields

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.google.sps.servlets;
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java]: This file isn't a servlet, so it should be under a different folder, e.g. data.

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  String text;
+  String timestamp;
+
+  public Comment(){
+  }
+
+  public Comment(long id, String text, String timestamp){
+    this.id = id;
+    this.text = text;
+    this.timestamp = timestamp;
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java]: I know that the Task example use the variable name timestamp, but that it doesn't have a object meaning, so I would suggest update it to something as 'creation_time' or a similar one that you would prefer.

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 2 above)
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Why are you choosing a String to represent timestamp? That is a completely valid option, but probably you'll have to do more casts. Instead you can use a long. Usually that is a best to way to represent timestamp when we communicate between services.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -11,38 +11,81 @@
...
+    response.getWriter().println(gson.toJson(comments));
   }
 
-  private String convertToJson(ArrayList<String> comments) {
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String comment = getComment(request);
+    response.setContentType("text/html");
+    String timestamp = new SimpleDateFormat("dd-MM-yyyy").format(new Date());
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: If you use a long, you can just do: ' long timestamp = System.currentTimeMillis();'

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -25,24 +25,43 @@
...
+    System.out.println("I convert to json");
     Gson gson = new Gson();
     String json = gson.toJson(comments);
     return json;
   }
+
+   /** Returns the comment from the comment box.*/
+   private String getComment(HttpServletRequest request) {
+     String commentString = request.getParameter("message");
+     System.out.println("I am here");
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Same about System.out.println

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -25,24 +25,43 @@
...
+
+     return commentString;
+   }
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:88]: Same comment that before, you can just return request.getParameter("message");

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -25,24 +25,43 @@
...
     String json = convertToJson(comments);
 
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    System.out.println("I do POST");
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Avoid keep System.out.println on the submitted code. They are a great help during development, but in the final code that avoid the code cleanness and readability.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -25,24 +25,43 @@
...
+    comments.add(comment);
+    response.setContentType("text/html");
+    response.getWriter().println("Comment submitted!");
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+  }
+
   private String convertToJson(ArrayList<String> comments) {
+    System.out.println("I convert to json");
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: System.out.println

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -25,24 +25,43 @@
...
+  ArrayList<String> comments = new ArrayList<String>();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> comments = new ArrayList<String>();
-    comments.add("Hello");
-    comments.add("How are you?");
-    comments.add("Greetings");
-
+    System.out.println("I do GET");
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: System.out.println

**Review by joanapascoal — CHANGES_REQUESTED** [review]
I really like the clean way that the code is organized!

Just adding a few comments that I believe that could improve it. Keep the great work!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed in [https://github.com/anamatasanch/SPS-portfolio/pull/12/commits/d3073c67f9f36eb9cc31a991921c1cd19f8475cc](url)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 9 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed in https://github.com/anamatasanch/SPS-portfolio/pull/12/commits/d3073c67f9f36eb9cc31a991921c1cd19f8475cc

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Will not happen again!

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 1 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 3 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 4 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Renamed it "datePosted"

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 2 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:88]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/Comment.java (hunk 2 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/Comment.java]: Fixed!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -11,38 +11,80 @@
...
+    response.getWriter().println(gson.toJson(comments));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String comment = getComment(request);
+    response.setContentType("text/html");
+    long datePosted = System.currentTimeMillis();
+    //String datePosted = new SimpleDateFormat("dd-MM-yyyy").format(new Date());
```
Comment (joanapascoal) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This comment can be removed.

**Review by joanapascoal — APPROVED** [review]
LGTM!

Thank you for updating :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (anamatasanch) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done!

**Pull request merged**